### PR TITLE
Acquire Token Silent Flow Refactor

### PR DESCRIPTION
**Pull request by RPangrle**
* Rename the ADTokenCacheAccessor protocol to ADTokenCacheDataStore
* Add a ADTokenCacheAccessor class for retrieving, adding and removing ADTokenCacheItems from a ADTokenCacheDataStore
* Remove the ADAuthenticationRequest+TokenCache category
* Move the functionality in the -requestWithServer: methods in ADAuthenticationRequest+WebRequest to ADWebAuthRequest, a subclass of ADWebRequest
* Move the AcquireTokenSilent flow into ADAcquireTokenSilentHandler

### CODE REVIEW

**Comment by RPangrle** [comment]
@jasoncoolmax @weijjia

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+                              resource:(NSString *)resource
+                              clientId:(NSString *)clientId
+                           redirectUri:(NSString *)redirectUri
+                            identifier:(ADUserIdentifier *)identifier
+                         correlationId:(NSUUID *)correlationId
+                            tokenCache:(ADTokenCacheAccessor *)tokenCache
+                       completionBlock:(ADAuthenticationCallback)completionBlock
+{
+    ADAcquireTokenSilentHandler* handler = [ADAcquireTokenSilentHandler new];
+    handler->_authority = authority;
```
Comment (jasoncoolmax) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: do we need to retain them? as later we will release them

Comment (RPangrle) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: Yes, we do.

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+        {
+            // If we failed to remove the item with an error, then return that error right away
+            completionBlock([ADAuthenticationResult resultFromError:error correlationId:_correlationId]);
+            return;
+        }
+        
+        if (!item.userInformation.userId)
+        {
+            // If we don't have any userInformation in this token that means it came from an authority
+            // that supports
```
Comment (jasoncoolmax) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: the sentence seems not complete yet :P

Comment (RPangrle) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: 👍

File: ADAL/src/request/ADAcquireTokenSilentHandler.m (hunk 1 above)
Comment (RPangrle) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: 👍

**Comment by jasoncoolmax** [comment]
I signed off :)

**Comment by RPangrle** [comment]
I'm going to close this pull request and reopen it against a different branch. I don't want to pull this into the 2.2 release as it's a bit too much code change for what is supposed to be a more narrowly focussed release. We'll save this for a 2.3

**Pull request closed without merging**